Added missing popup menu functionality

ckardaris opened the pull request:
The existing code had this functionality planned but not implemented.
I tried my best to make it work and I updated the README file with
instructions on how to use it.

ckardaris committed:
Added missing popup menu functionality

The existing code had this functionality planned but not implemented.
I tried my best to make it work and I updated the README file with
instructions on how to use it.

jonhoo commented:
Ah, excellent, thank you! I'll leave some in-line comments.
I think @bdantas might also be interested in this.

jonhoo commented on mktrayicon.c:
```diff
@@ -256,10 +300,99 @@ gpointer watch_fifo(gpointer argv)
+			if (param != NULL && *param == '\0') {
+#ifdef DEBUG
+				printf("Removing onmenu handler\n");
+#endif
```
I _think_ you have to `free(param)` here.

ckardaris replied:
I will add the `free(params)` where necessary in the next commit

jonhoo commented on mktrayicon.c:
```diff
@@ -256,10 +300,99 @@ gpointer watch_fifo(gpointer argv)
+			
+			menusize = 0;
+
+			if (param != NULL && *param == '\0') {
```
I wonder if it's ever the case that `param` is `NULL` here. If so, presumably we should still break?

ckardaris replied:
I believe since it comes from a `malloc` it can't be really `NULL` unless it fails. Maybe it's useful we add an `if` above the `switch` statement

jonhoo commented on mktrayicon.c:
```diff
@@ -256,10 +300,99 @@ gpointer watch_fifo(gpointer argv)
+				break;
+			}
+			if (param[0] == ',' || param[0] == '|'){
+				printf("Action string cannot begin with ',' or '|'. Use '\\' to escape\n");
```
Is there any particular reason to explicitly disallow this?

ckardaris replied:
The first character isn't anyway checked in the `while` loop below but for consistency I disallowed the `delimiters` to be used there. If you want to use them it's better to `escape` them. Makes the logic for the whole parameter the same.

jonhoo commented on mktrayicon.c:
```diff
@@ -256,10 +300,99 @@ gpointer watch_fifo(gpointer argv)
+			int commas = 0;
+			int bars = 0;
+			if (len < 3){
+				break;
```
I think this and all the other `break`s below have to `free(param)`

ckardaris replied:
You are right. I forgot to check this. Since we don't use it anymore we have to free.

jonhoo commented on mktrayicon.c:
```diff
@@ -256,10 +300,99 @@ gpointer watch_fifo(gpointer argv)
+				break;
+			}
+			if (param[len-2] == '|' && param[len-3] != '\\'){
+				printf("'|' can't be second of last. Use '\\' to escape\n");
```
Is there any particular reason to explicitly disallow this? Won't it just be equivalent to having no defined command for a label? Even so, it's probably easier to detect this _after_ parsing than before.

ckardaris replied:
Not really. It just made the logic clearer for me. Because one character after a valid `|` at the end has no point (a label without action)

jonhoo replied:
I think labels without an action is actually a valid use-case. For example, if you want the context menu to show some additional information beyond what is shown in the tooltip.

ckardaris replied:
I checked and it works for all cases except the last entry (because of the explicit condition to disallow `,` at the end). It generally accepts `<label>,|` as a sequence, creating an empty entry doing nothing. I will see how the parsing loops work if I remove that condition.

jonhoo commented on mktrayicon.c:
```diff
@@ -256,10 +300,99 @@ gpointer watch_fifo(gpointer argv)
+				break;
+			}
+			if (param[len-1] == ',' && param[len-2] != '\\'){
+				printf("',' can't be last. Use '\\' to escape\n");
```
Is there any particular reason to explicitly disallow this?

ckardaris replied:
Same as the other cases. For constistency only. Maybe we could allow for accepting a   label with no command. (Now that I think about it; I don't think I've checked this condition)

jonhoo commented on mktrayicon.c:
```diff
@@ -256,10 +300,99 @@ gpointer watch_fifo(gpointer argv)
+				break;
+			}
+			int i = 1; 
+			while (i < len - 1){
```
Is there any particular reason you want this to be a `while` instead of a `for`?

ckardaris replied:
Not really. Pick of the moment I guess. I will change it.

jonhoo commented on mktrayicon.c:
```diff
@@ -256,10 +300,99 @@ gpointer watch_fifo(gpointer argv)
+				i++;
+			}
+			if (commas > bars + 1 || bars > commas){
+				printf("Unbalanced ',' or '|'. Use '\\' to escape\n");
```
Isn't it easier to just detect this when parsing below instead?

ckardaris replied:
At first glance I thought your point made sense. But now that I checked it again, if I make the checks simultaneously while parsing, I may have to free already allocated memory for the `onmenu` array

jonhoo replied:
It probably makes sense to pre-count the number of non-escaped `,`, that is true :)

jonhoo commented on mktrayicon.c:
```diff
@@ -256,10 +300,99 @@ gpointer watch_fifo(gpointer argv)
+			menu = gtk_menu_new();
+			int last = -1;
+			int item = 0;
+			i = 1;
```
Shouldn't this start at `[0]`?

ckardaris replied:
If you mean `last` no because of the index arithmetic I use below. It's a bit confusing I know. I may have to clear that up. The same applies for `i` since I have forbidden the 0 index to be one of the delimiter characters from the code above, so no use in starting from 0. Anyway, I can look to clear that up.

jonhoo commented on mktrayicon.c:
```diff
@@ -256,10 +300,99 @@ gpointer watch_fifo(gpointer argv)
+			int last = -1;
+			int item = 0;
+			i = 1;
+			while(item < menusize && param[i] != '\0'){
```
I think I'd just write this as a nested `for` loop where the outer loop walks the `,`, and the inner loop find the `|`. Should make the code a little simpler I think.

ckardaris commented:
So to summarize:
- adding free(param) where necessary
- changing while loops where it can be a for loop
- i thought about revising the whole format of the parameter regarding the way to accept label-only entries and I think it should accept the following `<label1>|` or `<label>,|` if we need no action taken

jonhoo commented:
I wonder if we could use `strtok` to clean this up pretty significantly. Essentially, iterate over `strtok(',')`, taking care to ignore instances of `\,`. Should be pretty nice.

Another observation, and one we may want to ignore for the time being, is that people won't expect to have to escape `,` or `|` if it appears inside of a quoted string. If I write:
```
m "'foo,bar'|echo"
```
My expectation would be that the label comes out as `foo,bar`, with the action `echo`.

ckardaris committed:
Fixed accepted format of parameter for command 'm'

Added free() commands where appropriate
Made the logic a bit clearer and resistant to weird inputs

ckardaris commented:
So, I made some changes. Check them and tell me what we can do. I don't know if there are any bugs, because coding late is not the best, but I did a primitive test and it seems to be working. I will revise any errors tomorrow.

ckardaris commented:
> Another observation, and one we may want to ignore for the time being, is that people won't expect to have to escape `,` or `|` if it appears inside of a quoted string. If I write:
> 
> ```
> m "'foo,bar'|echo"
> ```
> 
> My expectation would be that the label comes out as `foo,bar`, with the action `echo`

This will be a bit trickier, but doable anyway. For now, though, I agree, we have to make the basic part work well.

jonhoo commented on mktrayicon.c:
```diff
@@ -320,79 +328,92 @@ gpointer watch_fifo(gpointer argv)
+			//we save so that we can free later even if we increase the pointer
+			
+			while (param[0] == ',' || param[0] == '|'){
+				// ignore delimiter characters in first spot
```
I'm not sure I follow why this is necessary? If the user puts `,` as the first character, then I think we should just treat that as any other empty label. Not clear that special-casing the _first_ one serves much of a purpose. Maybe _later_ filtering out any menu items with an empty label would be a better way to go?

ckardaris replied:
You have a point here. As I explain at the other comment below, I haven't made it to work with empty labels (meaning a blank spot at the menu, with or without action). Do we really need something line that?

But it works well, I think, with labels without action, an info spot at the menu, like you had suggested at the first commit.

jonhoo replied:
I think it's okay for us to simply ignore (and maybe warn) if we detect a menu item with an empty label. But I don't think we should do it in the matching up here (and thus only at the start/end). Instead, I think we should parse all the entries, and then warn/ignore on _any_ menu item with an empty label after we've parsed them.

jonhoo commented on mktrayicon.c:
```diff
@@ -320,79 +328,92 @@ gpointer watch_fifo(gpointer argv)
-				break;
+			while ((param[len-1] == ',' || param[len-1] == '|') && param[len-2] != '\\'){
+				// ignore delimiters at the end too
+				// it also helps dividing the string in entries -> bars+1
```
Same here, I don't think we should treat the beginning and end of the string specially.

jonhoo commented on mktrayicon.c:
```diff
@@ -320,79 +328,92 @@ gpointer watch_fifo(gpointer argv)
-					commas++;
-				else if (param[i] == '|' && param[i-1] != '\\')
+
+			// we can't accept 2 straight commas, as it becomes ambiguous
```
Why is it ambiguous? It just signifies an empty entry.

jonhoo commented on mktrayicon.c:
```diff
@@ -320,79 +328,92 @@ gpointer watch_fifo(gpointer argv)
 			
 			// Create the onmenu array which stores structs with name, action properties
-			menusize = commas;
+			menusize = bars + 1;
```
It seems odd to me to use `bars` here. Do we want to _require_ that every menu item has a command? Label-only menu items have some useful use-cases I think.

ckardaris replied:
Entries are separated by bars. So `label1|label2|label3` will create 3 empty labels

jonhoo replied:
Ah, sorry, I though `,` was the item separator and `|` was the label/action separator! My bad.

To me, `label,|` and `,|label` should both be accepted. They both parse into two menu items. The first produces a menu item with label "label" and no action, followed by a menu item with no label or action. The second produces a menu item with no label or action, followed by a menu item with the label "label" and no action. As mentioned earlier, when we actually _walk_ the menu items to produce the context menu, we could skip parsed items with no label and warn the user about them.

ckardaris replied:
So if I understand correctly you want to have `#entries = #bars + 1` whatever the position the bars are in

jonhoo replied:
Exactly.

ckardaris replied:
Ok, given the code structure it should be relatively easy to implement. I will make a new commit when it's ready.

ckardaris replied:
Quick question. What should we do with two straight commas? Currently it's disallowed. I think we should leave it at that

jonhoo replied:
Yeah, I think two commas in one menu item should be an error.

jonhoo commented:
I left some comments. Feel free to mark resolved comments as "resolved" in the GitHub UI :)

ckardaris commented:
To summarize. The code currently can:
- add label-action entry anywhere
- add label-only entry anywhere
- I believe it can handle empty entries (separators) anywhere except start and end (due to ignoring the chars); I have to check this one though. I will do it later.

If we want the ability to add separators at start and end, I should modify the code.

ckardaris committed:
Fix on NULL param for m command and freeable pointers

ckardaris committed:
Minor fix

ckardaris committed:
Accept empty entries everywhere

ckardaris commented:
UPDATE 2: Empty entries are accepted everywhere

Now, we should discuss how we want to handle the parameter when passed inside quotes

ckardaris commented:
I added the feature we were talking about. Now it is possible to surround string with single quotes and escape all characters inside (except the single quote character of course, which still needs escaping in that case). 

It seems to be working as expected (I tested some edges cases, but not all of course)

Only problem, as I explain in the commit message as well, is if the first label (and only that) is a label inside quotes. The problem goes away if I surround the whole parameter with quotes. I identified the most probable cause of this problem, being in the part of the inner loop handling the case of single quotes surrounding the parameter, but I didn't alter anything before discussing it a bit.

Code examples of what works and what not:
`echo "m Test1|'Test2,||,',xterm|Test3" > /tmp/fifo` will create 3 labels, 1st and 3rd doing nothing, and the 2nd being Test,||, and opening xterm, as expected
 `echo "m Test1|Test2',||,',xterm|Test3" > /tmp/fifo` will do the same (the quotes can be wherever inside a label and everything is processed correctly
`echo "m 'Test1',xterm" > /tmp/fifo` doesn't work
`echo "m 'Test1',xterm" > /tmp/fifo` will work as expected, since we surrounded the whole parameter in quotes

I haven't read thoroughly the way it reads the parameter in the inner loop, but I will do. 
In any case, if you can understand what goes wrong while parsing this and give a simple explanation to me it would help. It's a small bug, so no worries.

ckardaris commented:
Discussion topic: currently the quotes are parsed linearly allowing for only one level of surrounding strings. So `'aaa','bbb'` translates to `aaa` `,` `bbb`. Do we need it to be stack-like; first to open is last to close? So `'aaa','bbb'` translates to `aaa,bbb`

jonhoo commented:
I think single level is probably okay!

jonhoo commented on mktrayicon.c:
```diff
@@ -334,27 +365,53 @@ gpointer watch_fifo(gpointer argv)
+			 */
 			int straight = 0;
 			int bars = 0;
+			int qEven = 1; // if single quotes are even -> we don't count anything 
```
I would just call this `inQuoted`

ckardaris replied:
To not change the logic of the booleans, I will change it to outQuoted, which is the equivalent of qEven

jonhoo commented on mktrayicon.c:
```diff
@@ -19,12 +19,39 @@
+			 * so we move accordingly, copying everything
+			 */
+			i++;
+			while (src[i] != '\''){
```
I believe this needs a `&& src[i] != '\0'` clause

ckardaris replied:
At first glance it does. I thought about it while writing it. But to call this function we have made sure that number of quotes of `src` is even. So really, if we are in there it means that we will **surely** find the other quote to end the sequence. For safety, and to be consistent, we can add it though.

jonhoo commented on mktrayicon.c:
```diff
@@ -19,12 +19,39 @@
-	for (i = 0; i < n && src[i] != '\0'; i++){
-		if (src[i] != '\\'){
+	while (i < n && src[i] != '\0'){
+		if (src[i] == '\''){ 
```
I think you don't need this entire special-case. Instead, just remember if the first character was a `'`. Then, copy the string as before (skipping `\`). When you hit the end of the string (`\0`), if the first character was a `'`, check that the last character of `dest` is now a `'`. Then, turn it into a `\0`.

ckardaris replied:
The logic is that we can input an `src` that is not all in quotes and it will parse it correctly. So there are cases that the first char is not a quotes but there are quotes inside.
For example, `test'|||'end` will be parsed correctly, as one word `test|||end`

jonhoo replied:
I think it's totally fine for us to not support that, and only support whole-sale quoting of the argument. It'll make the code nicer without (I think) removing any particularly important use-cases.

ckardaris replied:
I think it's trickier (not much, but still) to remove this "behavior". What I can look into is changing the function and using this nice idea of remembering what we are reading. Much like the `outQuoted` (or `inQuoted`) variable. You are right that it is not so clean and readable now, so it should be written to be more intuitive

ckardaris commented:
I made some changes in the function. How do you see it now?

jonhoo commented on mktrayicon.c:
```diff
@@ -372,30 +372,30 @@ gpointer watch_fifo(gpointer argv)
 				int straight = 0;
 				int bars = 0;
-				int qEven = 1; // if single quotes are even -> we don't count anything 
+				int outQuoted = 1; // if single quotes are even -> we are outQuoted -> we can count 
```
I think `outQuoted` is a really hard-to-read variable name. My vote would be for inverting the booleans and calling it `inQuote` instead.

ckardaris replied:
I will change that

jonhoo commented on mktrayicon.c:
```diff
@@ -21,38 +21,32 @@ char* strncpy_esc(char *dest, const char *src, size_t n)
+				index++;
+				i+=2; 
+			}
+			else if (!outQuoted){ // non escaping '\' in quotes
```
This is weird to me. Why should `\` behave differently if you are inside or outside quotes?

ckardaris replied:
If you are inside quotes you don't escape any characters (except the quote characters if needed). So a `\` is a character you want to print. If you are outside you escape. That was the whole point with giving the option to quote arguments. Or have I misunderstood something?

jonhoo replied:
Ah, I was thinking of `'` as a way to use characters like `|` and `,` inside of a given label or action without having to escape them. I don't know whether or not it should also "turn off" `\` escaping for non-`'` characters.

Thinking some more about this, another option is to forego quoting altogether here, and _just_ provide `\` as a way to escape `\` and `,`. What do you think?

ckardaris replied:
Well, I guess we could leave it as is. Another option doesn't hurt anyone. And the escaping with just `\` works from the beginning, so it is ok.

jonhoo replied:
It would make the code diff a _lot_ smaller and nicer to read. Since _just_ `\` escaping is much less code, and sufficient to solve the problem, I'm starting to lean towards just doing that and removing the quoting code. How do you feel about that?

ckardaris replied:
I thought it was enough to begin with. So, we can revert back to the commit implementing just that. It will be a bit of a pain to check if I made any useful additions while implementing the "quote" code. Let me see what I can do

ckardaris replied:
Ok we are back to the commit before I added quote support. I made a copy of the quote code in the `quotes` branch just in case.

ckardaris replied:
I will have to make some changes in `master` now. I see some errors

ckardaris committed:
Fix in escape logic

ckardaris commented:
I guess you could start testing the feature a bit more thoroughly. I believe I have done a good job in that matter also, but you being an outside observer can maybe find some weird edge case.

jonhoo commented:
Excellent, this looks quite good to me now! Sorry for the back-and-forth, and thank you for the hard work. The only remaining thing I see is that I would like all the `if` arms to have curly brackets. The code currently omits them for single-statement arm bodies, such as:
```c
if (!params)
    break;
```
Which is just asking for trouble down the line. Once that's done (I think there were a couple of such `if`s), I'm happy to merge!

ckardaris committed:
Add curly boys in single-statement if/for blocks

ckardaris commented:
I believe it is done. Check again to make sure. I found some `for` statements missing the brackets, so I added them as well.

ckardaris commented:
And let me update the README real quick

jonhoo commented:
Looks good to me!

ckardaris committed:
Updated README

ckardaris commented:
I believe we are ready

jonhoo commented on README.md:
```diff
@@ -43,10 +43,15 @@ Quoted strings are terminated by a matching quote at the end of a line
+between entries (label+command).   
+If you want to use these 2 characters in a label or command, you have to escape 
+them with `\`.  
+If you want to have an entry with just a label and no command to be executed, you can ommit the 
```
"omit"

ckardaris replied:
Too many c**ommit**s

jonhoo commented on README.md:
```diff
@@ -43,10 +43,15 @@ Quoted strings are terminated by a matching quote at the end of a line
+them with `\`.  
+If you want to have an entry with just a label and no command to be executed, you can ommit the 
+`,<cmd>` part.  
+If you want an empty label, you can just add a second `|` delimiter after the previous one.  
```
I would write this as:
> If you want an empty label (e.g., as a separator), just add a second

jonhoo commented on README.md:
```diff
@@ -43,10 +43,15 @@ Quoted strings are terminated by a matching quote at the end of a line
-them with `\`.
-
-Example command: `echo "m Browser,firefox|Terminal,xterm" > /tmp/test` (where `mkfifo /tmp/test` has been executed before)
+between entries (label+command).   
```
I think this reads better if we don't have these additional line-breaks. Just keep it all as one paragraph with regular text-wrap linebreaks.

ckardaris committed:
Fix README

jonhoo commented on README.md:
```diff
@@ -23,6 +23,9 @@ are supported:
   - `c <cmnd>`: Set the command to be execute when the user clicks the icon 
 		(`cmnd` is passed to `/bin/sh -c`)
   - `c`: Remove the click handler
+  - `m <label1>,<cmd1>|<label2>,<cmd2>|...`: Set the labels and the corresponding commands to be executed when the user opens the icon menu (right-click usually) 
```
This line should be wrapped.

jonhoo commented on README.md:
```diff
@@ -39,6 +42,14 @@ Quoted strings are terminated by a matching quote at the end of a line
 to continue a quoted string, prefix it with a `\`.
 
+The m(enu) command uses `,` as a delimiter between label and command and `|` as a delimiter
+between entries (label+command).If you want to use these 2 characters in a label or command, you have to escape 
```
```suggestion
between entries (label+command). If you want to use these 2 characters in a label or command, you have to escape 
```

jonhoo commented on README.md:
```diff
@@ -39,6 +42,14 @@ Quoted strings are terminated by a matching quote at the end of a line
 (ignoring whitespace). To escape a quote character at the end of a line
 to continue a quoted string, prefix it with a `\`.
 
+The m(enu) command uses `,` as a delimiter between label and command and `|` as a delimiter
```
I think the text wrapping here is wider than for the other lines in the file?

jonhoo commented:
I went ahead and merged this -- thanks again! I'll do some minor tweaking to styling afterwards, but this is a good change :)

ckardaris commented:
Glad I could help. I was really excited when I first saw the project, cause I was looking for a way to make this on my own. So, it's nice to know that I have contributed a bit to it now.
